### PR TITLE
feat: implement the baseline node sdk

### LINKED ISSUE #2: [GLCC] GitLink Code Camp
The issue is about the first GitLink Code Camp activity.
Related link: https://mp.weixin.qq.com/s/upVI-2TQ0ezQGpzhDiF3Pg

## Requirement

We need a server-side Node.js SDK, which can work as other FeatureProbe server side SDKs, take Java SDK for example.

You need to be familiar with Typescript, have knowledge of network and multi-thread programming.

## Expectation

The Node.js SDK should work as Java SDK, including Targeting calculation and metrics report, and should pass all the acceptance tests including performance test.

## Contact

@frankzhaopeng 
email: poryzhao@featureprobe.com

### PR DESCRIPTION
**Pull request by hezean**
Closes #2

### CODE REVIEW

File: example/index.html [hunk 1]
```diff
@@ -0,0 +1,106 @@
+<!DOCTYPE html>
```
Comment (frankzhaopeng) [inline example/index.html]: Delete this index.html, Add some demo js instead

File: package.json [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+    "rollup-plugin-serve": "^1.1.0",
+    "rollup-plugin-typescript2": "^0.31.2",
+    "ts-jest": "^28.0.8",
+    "typescript": "^4.7.4"
+  },
+  "dependencies": {
+    "@types/semver": "^7.3.12",
+    "isomorphic-fetch": "^3.0.0",
+    "ts-node": "^10.9.1",
+    "whatwg-fetch": "^3.6.2"
```
Comment (frankzhaopeng) [inline package.json]: Keep isomorphic-fetch, remove whatwg-fetch in this file

File: rollup.config.js [hunk 3]
```diff
@@ -0,0 +1,31 @@
+import typescript from "rollup-plugin-typescript2";
+import minify from "rollup-plugin-babel-minify";
+import resolve from "rollup-plugin-node-resolve";
+import commonjs from "rollup-plugin-commonjs";
+import nodePolyfills from "rollup-plugin-node-polyfills";
+import json from "@rollup/plugin-json";
+
```
Comment (frankzhaopeng) [inline rollup.config.js:8]: Add rollup-plugin-node-builtins plugin to solve building crypto error

File: src/Event.ts [hunk 4]
```diff
@@ -0,0 +1,216 @@
+// "use strict";
+
+import "whatwg-fetch";
```
Comment (frankzhaopeng) [inline src/Event.ts]: Remove whatwg-fetch, can not be used in Node.js

File: src/Evaluate.ts [hunk 5]
```diff
@@ -0,0 +1,616 @@
+"use strict";
+
+import { createHash } from "crypto";
```
Comment (frankzhaopeng) [inline src/Evaluate.ts]: temporary solution:

/* eslint-disable-next-line */
const SemVer = require("semver/classes/semver");

File: src/Evaluate.ts [hunk 6]
```diff
@@ -0,0 +1,616 @@
...
+  get split(): Split | null {
+    return this._split;
+  }
+
+  set split(value: Split | null) {
+    this._split = value;
+  }
+
+  public evalIndex(user: FPUser, toggleKey: string): IHitResult {
+    if (this._select !== null) {
```
Comment (frankzhaopeng) [inline src/Evaluate.ts]: this._select may be undefined

File: test/Event.test.ts [hunk 7]
```diff
@@ -0,0 +1,36 @@
+import fetchMock from "jest-fetch-mock";
```
Comment (frankzhaopeng) [inline test/Event.test.ts]: Node.js SDK implementation should adhere to these specs in this repo:https://github.com/FeatureProbe/server-sdk-specification

File: package.json [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    "prettier": "^2.7.1",
+    "rollup": "^2.70.2",
+    "rollup-plugin-babel-minify": "^10.0.0",
+    "rollup-plugin-commonjs": "^10.1.0",
+    "rollup-plugin-node-polyfills": "^0.2.1",
+    "rollup-plugin-node-resolve": "^5.2.0",
+    "rollup-plugin-serve": "^1.1.0",
+    "rollup-plugin-typescript2": "^0.31.2",
+    "ts-jest": "^28.0.8",
+    "typescript": "^4.7.4"
```
Comment (frankzhaopeng) [inline package.json:65]: Add typedoc as devDependency, by executing `npm run doc` to generate output
We will build those static assets later on our website

File: src/Sync.ts [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+import "whatwg-fetch";
+// import { TinyEmitter } from "tiny-emitter";
+// import { pino } from "pino";
+
+import { Repository } from "./Evaluate";
+
+import pkg from "../package.json";
+
+const UA = `Node/${pkg.version}`;
+// const logger = pino({
```
Comment (frankzhaopeng) [inline src/Sync.ts]: logs are needed in Node.js SDK as well

Comment (hezean) [reply src/Sync.ts]: using `pino`, since it's high-performance and flexible to transport

File: src/Event.ts [hunk 10]
```diff
@@ -0,0 +1,214 @@
+'use strict';
+
+import pino from 'pino';
+
+require('isomorphic-fetch');
+import pkg from '../package.json';
+
+const UA = `Node/${pkg.version}`;
```
Comment (frankzhaopeng) [inline src/Event.ts:8]: Use require to make comman `npm run doc` available

const pkg = require('./package.json');

File: typedoc.js [hunk 11]
```diff
@@ -0,0 +1,9 @@
+import pkg from "./package.json";
+const PKG_VERSION = pkg.version;
+
+module.exports = {
+  out: "./doc",
```
Comment (frankzhaopeng) [inline typedoc.js]: change the output folder to `./docs`, and add this folder to repository

**Review by hezean — COMMENTED** [review]
LGTM

File: src/Evaluate.ts [hunk 12]
```diff
@@ -0,0 +1,598 @@
...
+        warning = hitResult.reason;
+      }
+    }
+    return this.defaultResult(user, this._key, defaultValue, warning);
+  }
+
+  private hitValue(hitResult: IHitResult | undefined, defaultValue: any, ruleIndex?: number): FPToggleDetail {
+    const res = {
+      value: defaultValue,
+      ruleIndex: ruleIndex || null,
```
Comment (frankzhaopeng) [inline src/Evaluate.ts]: if ruleIndex is 0, expression 0 || null will return null, actually we need 0 not null.
same with line 187

File: package.json [hunk 13]
```diff
@@ -0,0 +1,62 @@
+{
```
Comment (frankzhaopeng) [inline package.json:1]: add a new fields: files to include dist folder, files in this folder will be published in npm packages.
files: [
"dist"
]

File: package.json [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+  "name": "server-sdk-node",
```
Comment (frankzhaopeng) [inline package.json]: change the package name to "featureprobe-server-sdk-node"

**Pull request merged**